pg-core/src/clj/pg/core.clj

```diff
@@ -1242,81 +845,61 @@
 
 
 ;;
-;; Connectable source abstraction
+;; Pool?
 ;;
 
-(defprotocol IConnectable
-
-  (-borrow-connection [this])
-
-  (-return-connection [this conn]))
-
-
-(extend-protocol IConnectable
-
-  Connection
-
-  (-borrow-connection [^Connection this]
-    this)
-
-  (-return-connection [this ^Connection conn]
-    nil)
-
-  Pool
-
-  (-borrow-connection [this]
-    (.borrowConnection this))
-
-  (-return-connection [this ^Connection conn]
-    (.returnConnection this conn))
-
-  IPersistentMap
-
-  (-borrow-connection [this]
-    (connect this))
-
-  (-return-connection [this ^Connection conn]
-    (close conn))
-
-  Config
+(defn pool?
+  "
+  True if a value is a Pool instance.
+  "
+  [x]
+  (instance? Pool x))
 
-  (-borrow-connection [this]
-    (Connection/connect this))
 
-  (-return-connection [this ^Connection conn]
-    (close conn))
+(defn pool
+  "
+  Run a new Pool from a config map or a URI.
+  "
+  (^Pool [src]
+   (-> src
+       (src/-to-config)
+       (Pool/create)))
 
-  Object
+  (^Pool [^String host ^Integer port ^String user ^String password ^String database]
+   (Pool/create host port user password database)))
 
-  (-borrow-connection [this]
-    (error! "Unsupported connection source: %s" this))
 
-  (-return-connection [this ^Connection conn]
-    (error! "Unsupported connection source: %s" this))
+(defmacro with-pool
+  "
+  Execute the body while the `bind` symbol is bound
+  to a new Pool instance. Close the pool afterwards.
+  "
+  [[bind config] & body]
+  `(with-open [~bind (pool ~config)]
+     ~@body))
 
-  nil
 
-  (-borrow-connection [this]
-    (error! "Connection source cannot be null"))
+;;
+;; DEPRECATED
+;;
 
-  (-return-connection [this ^Connection conn]
-    (error! "Connection source cannot be null")))
+(defn ^:deprecated get-error-fields
+  "
+  Get a map of error fields from an instance of
+  `PGErrorResponse`. **Deprecated**, use `ex-data`
+  instead.
+  "
+  [^PGErrorResponse e]
+  (ex-data e))
 
 
-(defmacro on-connection
+(defmacro ^:deprecated with-tx
```

Comment:
The `with-tx` macro was designed such that it doesn't accept a binding symbol, and now causes issues. There is a new `with-transaction` macro below.